### ISSUE TITLE
Create Python API for querying splunk

### ISSUE THREAD

**Issue by EyalBerger** [hasadna/open-bus#177]
It will be much more easy for data analyses to set a connection between python and Splunk DB.

@cjer @AvivSela What do you think about it? does is possible?

I found this [doc](http://dev.splunk.com/view/python-sdk/SP-CAAAEE4).

May I try to implement it using my Splunk user? If one of you already started working on this kind of connection please share your code.

**Comment by AvivSela** [comment]
we planning to move the data outside of splunk, so when we thought about it we proffered to  export manually the data. If you think it will make things more easy in the meantime to have this workaround so go for it.

**Comment by EyalBerger** [comment]
Thanks. I will try to set this connection for us.

I downloaded splunk-sdk and tried to set a connection to our Splunk enterprise, using the Client class, but the connection has failed, throwing this error:

`TimeoutError: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond`
  
My guess is that I didn't passed the correct credentials - 

* HOST (**35.192.116.75 ?**)
* PORT (**8089 ?**)
* USERNAME (my user name)
* PASSWORD (my password)

@cjer @AvivSela do you know what are the correct credentials to set a connection?

**Comment by cjer** [comment]
@EyalBerger as for your last error, this has to do with network settings for the machine, not splunk credentials. I'll open the port.
As for the general issue. As I said, since splunk is our current database, there's no reason not to use it as best as we can for what we're trying to do. I think we should define the API we want to have for data science purposes, and then splunk will be our current (temporary?) backend. Maybe start with what's in the `Data Exporter` dashboard for basic raw data fetching (SIRI and GTFS is there).
P.S I won't be at the meetup tomorrow so please keep on updating the issue, and I'll try to help.

**Comment by AvivSela** [comment]
I open the port and succeed to connect to it with telnet

**Comment by EyalBerger** [comment]
Thanks, Now I get this error `HTTPError: HTTP 401 Unauthorized -- Login failed`.

Maybe I have to connect as an admin?

**Comment by cjer** [comment]
Your user should have REST capabilities. Can you paste your code here? (excluding credentials of course)

**Comment by EyalBerger** [comment]
`!pip install splunk-sdk`
`import sys`
`import getpass`
`sys.path.append('splunk-sdk-python-1.6.6')`
`import splunklib.client as client`

`# Create a Service instance and log in `
`service = client.connect(`
`    host=HOST,`
`    port=PORT,`
`    #schema = 'https', #????`
`    username=USERNAME,`
`    password=PASSWORD)
`
`# Print installed apps to the console to verify login`
`for app in service.apps:`
`    print (app.name)`

As for credentials, I used the HOST and POST I mentioned above, and for USERNAME and PASSWORD I passed my own username and password.

**Comment by cjer** [comment]
I just tried it with a test user I created which has the same credentials as you and it works perfectly.  Maybe there's a problem with the username/password you entered?

**Comment by EyalBerger** [comment]
Yes, oops...
I entered the my old password instead of my new one.
Now its works.
Thanks

**Comment by EyalBerger** [comment]
OK,
I developed a **read_siri** function which query the Splunk DB from python, but during working on some last issues, its raised this error:
`HTTPError: HTTP 503 Service Unavailable -- reason="Search not executed: The maximum number of concurrent historical searches for this user based on their role quota has been reached.", concurrency_category="historical", concurrency_context="user_instance-wide", current_concurrency=3, concurrency_limit=3`   

After a few minutes, its suddenly returned to work properly and didn't raised any error when querying the DB.

I assume that for daily working with my function we will need to update my *maximum number of concurrent historical searches*

My code:
* Set query_kwargs - 
 ```
query_kwargs = {'agency_id': 3,
                'route_id': 5189,
                'route_short_name': 501,
                #'time_recorded': '00:03:00',
               'max_columns': 100}

```
* Define **splunk_query_builder** function -
```
 def splunk_query_builder(**query_kwargs):
    query = 'search *'
        
    for param in query_kwargs:
        if query_kwargs[param]:
            if param != 'max_columns':
                query = query + ' {}={}'.format(param, query_kwargs[param])
            elif param == 'max_columns':
                query = query + ' | head {}'.format( query_kwargs[param])   
    
    return query
```

* Define **read_siri** function - 
```
def read_siri(query, username, password, host, port):
    
    print('started')
    
    #imports
    import sys
    import re
    import pandas as pd
    
    sys.path.append('splunk-sdk-python-1.6.6')
    import splunklib.client as client
    
    #connect to splunk
    service = client.connect(host=host, port=port, username=username, password=password)
    print('connection succeed')

    #query splunk
    oneshotsearch_results = service.jobs.oneshot(query)
    print('query succeed')
    
    #read results 
    results = oneshotsearch_results.read()
    print('read results succeed')
    
    #transform results to DataFrame
    columns = ['timestamp','agency_id','route_id','route_short_name',
           'trip_id_to_date','planned_start_time','bus_id',
           'time_recorded','predicted_end_time','lat','lon','date']

    siri_df = pd.DataFrame(columns=columns)
    
    results_rows = re.findall(b"trunc='0'>(.+?)<", results)
    
    for row in results_rows:
        data = row.decode('utf-8').split(',')
        siri_df.loc[len(siri_df)] = data
    
    print('finished')
    
    return siri_df
```

* Read_siri - 

```
siri_data = read_siri(splunk_query_builder(**query_kwargs),
          host =  HOST,
          port = PORT,
          username = USERNAME,
          password = PASSWORD)

siri_data.head()
```

**Comment by EyalBerger** [comment]
As for the last error I mentioned, I now use another query method that (still?) didn't raised this error, so for now no changes in configurations need to be made.

**Comment by EyalBerger** [comment]
The previous error (`503`) raised again and for few hours I can't run queries neither by code nor by the Splunk web app. Its will be great if someone could solve it and update `my maximum number of concurrent historical searches`.
As for this issue main subject, for now my pythonic **read_siri** function works very slow. I am working on improving it in two aspects- 
* Choosing the best Splunk API query method (there are about 4 options).
* Optimize my **Search Processing Language (spl)** query (I will open new issue under that topic).  

I am also intend to adjust the function to read also GTFS data from Splunk (*route_stats*).

**Comment by cjer** [comment]
I don't think the solution is updating maximum concurrent searches. The problem is that you are running very slow and long searches, and run them one after the other without cancelling the previous ones. Increasing concurrent searches will just hurt the server.
I think the best way to go forward is that we sit the coming Monday, if you're coming, and I'll tell you a bit about splunk optimization.
Just a quick remark - I see you're not giving a time range and an index to the the search. Your searches are searching textually on ALL available indexes, this will take a VERY long time.

**Comment by EyalBerger** [comment]
Yes, I think its will be great.  
Thanks for the tips.

**Comment by EyalBerger** [comment]
Hi, thanks to Dan I improved my SPL skills, and finished a first version of **read_splunk** function. 

I will handle documentation issues next week and share it.

This version have some known shortcomings for now, but I assume it could be still useful for data analyze.

The main challenge is to overcame the Splunk API limit of getting maximum of 50,000 events by one query.  As far as I understood, admin can change this limitation (see [here](https://docs.splunk.com/Documentation/Splunk/latest/Admin/Limitsconf?utm_source=answers&utm_medium=in-answer&utm_term=limits.conf&utm_campaign=refdoc) in [searchresults]), by its not recommended.
As for now, I am using a manual `for` loop for getting more results.  e.g: Its took my loop 15 minutes to read all Metropoline (agency id=15) SIRI data from the last 3 days (1400K rows).

**Comment by cjer** [comment]
Great!
Do you have a tutorial/example notebook that uses your API?
About getting more than 50k results besides changing `limits.conf`, I remember there are a few ways to do it. One was to create a job (instead of running a `oneshot`), and then exporting the results once it is finished (this [answer](https://answers.splunk.com/answers/39243/python-sdk-results-limited-to-50-000.html))

**Comment by EyalBerger** [comment]
I am working on it :)

As for getting more than 50k  results - I will work on it after I'll finish the first version, thanks!

**Comment by EyalBerger** [comment]
I uploaded a first version of SplunkUtills. **It's ready to merge now**.

Note that I compared my API results (when using a loop) to the Splunk app results, and founded a gap regarding the last day of the search (I send both queries in 11.5.2019, aiming to get last 30 days data - only the app returned full data from 11.4.2019).

After reviewing and merging this version, I will close this issue and will open a new one for some future improvements that need to be done in SplunkUtills.

**Comment by AvivSela** [comment]
could you elaborate about how much time it take to get the data from Splunk?
- a specific ride from yesterday
- all rides of specific route from yesterday
- a specific ride from 14/2/19
- all rides of specific route from 14/2/19

**Comment by EyalBerger** [comment]
* a specific ride from yesterday = There is no data in Splunk from yesterday. I queried for `search index=siri earliest=-8d latest=-7d agency_id=3 route_id=12405 planned_start_time=20:20:00`. 96 rows, 6 seconds.

* all rides of specific route from yesterday = There is no data in Splunk from yesterday. I queried for `search index=siri earliest=-8d latest=-7d agency_id=3 route_id=12405 `. 9009 rows, 6 seconds.

* a specific ride from 14/2/19 = I queried for ` search index=siri earliest="02/14/2019:00:00:00" latest="02/14/2019:23:59:59" agency_id=3 route_id=12405 planned_start_time=20:20:00`. 34 rows, 14 seconds.

* all rides of specific route from 14/2/19 - I queried for  `search index=siri earliest="02/14/2019:00:00:00" latest="02/14/2019:23:59:59" agency_id=3 route_id=12405`. 901 rows, 8 seconds.

**Comment by EyalBerger** [comment]
SplunkUtills had merged  - [SplunkUtills](https://github.com/hasadna/open-bus/tree/master/splunk/SplunkUtills). I am closing this issue.

I opened a new issue for some open tasks regarding SplunkUtills - #234.
